core/src/test/java/com/linecorp/armeria/client/HttpClientIntegrationTest.java

```diff
@@ -717,6 +736,59 @@ void testUpgradeRequestExecutesLogicOnlyOnce() throws Exception {
         clientFactory.close();
     }
 
+    @ParameterizedTest
+    @CsvSource({
+            "H1C, true", "H1C, false",
+            "H2C, true", "H2C, false"
+    })
+    void testResponseTimeoutHandler(SessionProtocol protocol, boolean useResponseTimeoutHandler) {
+        final AtomicReference<RequestLog> logHolder = new AtomicReference<>();
+        final IllegalStateException reqCause = new IllegalStateException("abort request");
+        final IllegalStateException logCause = new IllegalStateException("log cause");
+        final AtomicBoolean invokeResponseTimeoutHandler = new AtomicBoolean(false);
+        final HttpClient client = new HttpClientBuilder(server.uri(protocol, "/"))
+                .responseTimeout(Duration.ofSeconds(2))
+                .decorator((delegate, ctx, req) -> {
+                    if (useResponseTimeoutHandler) {
+                        ctx.setResponseTimeoutHandler(() -> {
+                            ctx.request().abort(reqCause);
+                            invokeResponseTimeoutHandler.set(true);
+                        });
+                    }
+                    logHolder.set(ctx.log());
+                    return delegate.execute(ctx, req);
+                })
+                .build();
+
+        final HttpRequestWriter writer = HttpRequest.streaming(HttpMethod.POST, "/slow");
+        final HttpResponse response = client.execute(writer);
+        await().untilAsserted(() -> {
+            assertThat(logHolder.get().isAvailable(RequestLogAvailability.COMPLETE)).isTrue();
+        });
+
+        if (useResponseTimeoutHandler) {
+            assertThat(invokeResponseTimeoutHandler).isTrue();
+            assertThatThrownBy(() -> response.aggregate().join()).isInstanceOf(CompletionException.class)
+                    .hasCauseInstanceOf(ClosedSessionException.class);
```

Comment:
I'm not sure if emitting `ClosedSessionException` is right or not. I think `ResponseTimeoutException` is better than `ClosedSessionException`.
And I think `reqCause` is slightly better than `ResponseTimeoutException`.
WDYT?

Reply:
Now, it returns `reqCause` instead of `ClosedSessionExcepiton`.